test/__init__.py

```diff
@@ -195,6 +195,7 @@ def __del__(self):
                     common.EVENTS_QUEUE_FREQUENCY,
                     self._stack))
             self.disable()
+            raise Exception(msg)
```

Comment:
This change avoids this crytpic error which happened at shutdown:
```
 [2021/06/21 21:18:28.948] Exception ignored in: <bound method client_knobs.__del__ of <test.client_knobs object at 0x7f694ea88b00>>
 [2021/06/21 21:18:28.948] Traceback (most recent call last):
 [2021/06/21 21:18:28.948]   File "/data/mci/039c3bbed0124de319c623e31b57b27e/src/test/__init__.py", line 196, in __del__
 [2021/06/21 21:18:28.948] ValueError: underlying buffer has been detached
```
https://spruce.mongodb.com/task/mongo_python_driver_tests_python_version_amazon1_test_ssl__platform~awslinux_auth_ssl~noauth_nossl_python_version~3.6_coverage~coverage_test_4.0_sharded_cluster_patch_3de63373aadc67518868b1db15b40e7bc8e981ab_60d0fca1c9ec44224bb1a9d1_21_06_21_20_54_58/logs?execution=0

If it happens again we should be able to see where the offending client_knobs was created.

Reply:
Once 3.12 is out we can actually get rid of our homebrewed client_knobs helper and instead use `unittest.mock`...

Reply:
If mock can do everything client_knobs does, including the warning message when it's not unpatched, then great!

Reply:
Found the culprit:
```
 [2021/06/22 23:40:22.788] Exception ignored in: <bound method client_knobs.__del__ of <test.client_knobs object at 0x7f074ac19f28>>
 [2021/06/22 23:40:22.788] Traceback (most recent call last):
 [2021/06/22 23:40:22.788]   File "/data/mci/e00bc7a4a5010873aa836e890f7b7fef/src/test/__init__.py", line 198, in __del__
 [2021/06/22 23:40:22.788] Exception:
 [2021/06/22 23:40:22.788] ERROR: client_knobs still enabled! HEARTBEAT_FREQUENCY=0.1, MIN_HEARTBEAT_INTERVAL=0.1, KILL_CURSOR_FREQUENCY=1, EVENTS_QUEUE_FREQUENCY=1, stack:
 [2021/06/22 23:40:22.788]   File "setup.py", line 351, in <module>
 [2021/06/22 23:40:22.788]     **extra_opts
 [2021/06/22 23:40:22.788]   File "/opt/python/3.6/lib/python3.6/site-packages/setuptools/__init__.py", line 153, in setup
 [2021/06/22 23:40:22.788]     return distutils.core.setup(**attrs)
 [2021/06/22 23:40:22.788]   File "/opt/python/3.6/lib/python3.6/distutils/core.py", line 148, in setup
 [2021/06/22 23:40:22.788]     dist.run_commands()
 [2021/06/22 23:40:22.788]   File "/opt/python/3.6/lib/python3.6/distutils/dist.py", line 955, in run_commands
 [2021/06/22 23:40:22.788]     self.run_command(cmd)
 [2021/06/22 23:40:22.788]   File "/opt/python/3.6/lib/python3.6/distutils/dist.py", line 974, in run_command
 [2021/06/22 23:40:22.788]     cmd_obj.run()
 [2021/06/22 23:40:22.788]   File "setup.py", line 123, in run
 [2021/06/22 23:40:22.788]     result = runner.run(suite)
 [2021/06/22 23:40:22.788]   File "/data/mci/e00bc7a4a5010873aa836e890f7b7fef/src/test/__init__.py", line 976, in run
 [2021/06/22 23:40:22.788]     result = super(PymongoXMLTestRunner, self).run(test)
 [2021/06/22 23:40:22.788]   File "/opt/python/3.6/lib/python3.6/site-packages/xmlrunner/runner.py", line 66, in run
 [2021/06/22 23:40:22.788]     test(result)
 [2021/06/22 23:40:22.788]   File "/opt/python/3.6/lib/python3.6/unittest/suite.py", line 84, in __call__
 [2021/06/22 23:40:22.788]     return self.run(*args, **kwds)
 [2021/06/22 23:40:22.788]   File "/opt/python/3.6/lib/python3.6/unittest/suite.py", line 114, in run
 [2021/06/22 23:40:22.788]     self._handleClassSetUp(test, result)
 [2021/06/22 23:40:22.788]   File "/opt/python/3.6/lib/python3.6/unittest/suite.py", line 163, in _handleClassSetUp
 [2021/06/22 23:40:22.788]     setUpClass()
 [2021/06/22 23:40:22.788]   File "/data/mci/e00bc7a4a5010873aa836e890f7b7fef/src/test/__init__.py", line 513, in wrap
 [2021/06/22 23:40:22.788]     return f(*args, **kwargs)
 [2021/06/22 23:40:22.788]   File "/data/mci/e00bc7a4a5010873aa836e890f7b7fef/src/test/__init__.py", line 513, in wrap
 [2021/06/22 23:40:22.788]     return f(*args, **kwargs)
 [2021/06/22 23:40:22.788]   File "/data/mci/e00bc7a4a5010873aa836e890f7b7fef/src/test/test_retryable_reads.py", line 60, in setUpClass
 [2021/06/22 23:40:22.788]     super(TestSpec, cls).setUpClass()
 [2021/06/22 23:40:22.788]   File "/data/mci/e00bc7a4a5010873aa836e890f7b7fef/src/test/utils_spec_runner.py", line 99, in setUpClass
 [2021/06/22 23:40:22.788]     cls.knobs.enable()
 [2021/06/22 23:40:22.788]   File "/data/mci/e00bc7a4a5010873aa836e890f7b7fef/src/test/__init__.py", line 171, in enable
 [2021/06/22 23:40:22.788]     self._stack = ''.join(traceback.format_stack())
```